incubator/validator-avro/src/main/java/io/aklivity/zilla/runtime/validator/avro/AvroValidator.java

```diff
@@ -14,132 +14,99 @@
+        this.format = config.format;
+        this.subject = catalog != null && catalog.subject != null
+                ? catalog.subject
+                : config.subject;
+        this.cache = new Int2ObjectCache<>(1, 1024, i -> {});
     }
 
-    @Override
-    public boolean read(
-        DirectBuffer data,
-        int index,
-        int length)
+    protected Schema fetchSchema(
+        int schemaId)
     {
-        boolean status = false;
-        byte[] payloadBytes = new byte[length];
-        data.getBytes(0, payloadBytes);
-        ByteBuffer byteBuf = ByteBuffer.wrap(payloadBytes);
+        Schema schema = null;
 
-        if (byteBuf.get() == MAGIC_BYTE)
+        if (schemaId == 0)
         {
-            int schemaId = byteBuf.getInt();
-            int valLength = length - 1 - 4;
-            byte[] valBytes = new byte[valLength];
-            data.getBytes(length - valLength, valBytes);
-
-            String schema = handler.resolve(schemaId);
-
-            if (schema != null && validate(schema, valBytes))
-            {
-                status = true;
-            }
+            schemaId = handler.resolve(subject, catalog.version);
         }
-        return status;
-    }
 
-    @Override
-    public boolean write(
-        DirectBuffer data,
-        int index,
-        int length)
-    {
-        boolean status = false;
-        String schema = null;
-        int schemaId = catalog != null ? catalog.id : 0;
-
-        byte[] payloadBytes = new byte[length];
-        data.getBytes(0, payloadBytes);
-
-        if (schemaId > 0)
+        if (cache.containsKey(schemaId))
         {
-            schema = handler.resolve(schemaId);
+            schema = cache.get(schemaId);
         }
-        else if (catalog != null && "topic".equals(catalog.strategy))
+        else
         {
-            schemaId = handler.resolve(subject, catalog.version);
-            if (schemaId > 0)
+            String schemaStr = handler.resolve(schemaId);
+            if (schemaStr != null)
             {
-                schema = handler.resolve(schemaId);
+                schema = new Schema.Parser().parse(schemaStr);
+                cache.put(schemaId, schema);
             }
         }
 
-        if (schema != null && validate(schema, payloadBytes))
-        {
-            status = true;
-        }
-
-        return status;
+        return schema;
     }
 
-    private boolean validate(
-        String schema,
-        byte[] payloadBytes)
+    protected boolean validate(
+        Schema schema,
+        byte[] bytes,
+        int offset,
+        int length)
     {
         boolean status = false;
         try
         {
-            reader = new GenericDatumReader(parser.parse(schema));
-            reader.read(null, decoder.binaryDecoder(payloadBytes, null));
+            reader = new GenericDatumReader<>(schema);
+            reader.read(null, decoder.binaryDecoder(bytes, offset, length, null));
```

Comment:
Use the same optimizations here as read validator for `reader` and `decoderFactory` /  `decoder`.